src/Lecture2.hs

```diff
@@ -253,15 +498,30 @@ Normally, this would be a sum type with several constructors
 describing all possible errors. But we have only one error in our
 evaluation process.
 -}
-data EvalError
-    = VariableNotFound String
-    deriving (Show, Eq)
+data EvalError 
+    = VariableNotFound String 
+    | MultipleVariablesFound String
+    | IntOverflow Int Int 
+    deriving (Show, Eq) 
 
 {- | Having all this set up, we can finally implement an evaluation function.
 It returns either a successful evaluation result or an error.
 -}
-eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval :: Variables -> Expr -> Either EvalError Int 
+eval _ (Lit n) = Right n
+eval vars (Var var) = case lookupAll var vars of
+    [] -> Left $ VariableNotFound var
+    x : [] -> Right x
+    _ : _ : _ -> Left $ MultipleVariablesFound var
+eval vars (Add a b) = case (a', b') of
+    (Left ae, _) -> Left ae
+    (_, Left be) -> Left be
+    (Right a'', Right b'') -> case safeSum a'' b'' of
+        Just res -> Right res
+        Nothing -> Left $ IntOverflow a'' b''
```

Comment:
Great implementation! 
Once you learn about monads, you'll be able to write this shorter ✂️